doc/apt-config.8
.TH "APT-CONFIG" 8 "14 Jun 2006" "APT-RPM" "apt-config"
.SH NAME
apt-config - APT Configuration Query program

.SH SYNOPSIS
\fBapt-config\fR [\fBoptions\fR] [\fB-o\fR \fIconfig\fR=\fIstring\fR] [\fB-c\fR=\fIfile\fR] \fBcommand\fR

.SH "DESCRIPTION"
\fBapt-config\fR is an internal program used by various portions of the APT
suite to provide consistent configurability.  It accesses the main
configuration file \fI/etc/apt/apt.conf\fR(5) in a manner that is easy to use
by scripted applications.

.SH "COMMANDS"
Unless the \fB-h\fR, or \fB--help\fR option is given, one of the commands
below must be present.
.TP
\fBshell\fR
Used to access the configuration information from a shell script.  It is
given pairs of arguments, the first being a shell variable and the second
the configuration value to query.  As output it lists a series of shell
assignments commands for each present value. In a shell script it should be
used like:
.IP
.nf
OPTS="-f"
RES=`apt-config shell OPTS MyApp::Options`
eval $RES
.fi
.IP
This will set the shell environment variable $OPTS to the value of
\fIMyApp::Options\fR with a default of \fB-f\fR. .IP The configuration item
may be postfixed with a /[fdbi].  \fBf\fR returns file names, \fBd\fR
returns directories, \fBb\fR returns true or false and \fBi\fR returns an
integer.  Each of the returns is normalized and verified internally.

.TP
\fBdump\fR
Shows the contents of the configuration space.

.SH "OPTIONS"
All command-line options may be set using the configuration file, the
descriptions indicate the configuration option to set.  For boolean
options you can override the config file by using something like 
\fB-f-\fR,\fB--no-f\fR, \fB-f=no\fR or several other variations.

.TP
\fB-h\fR, \fB--help\fR
Show a short usage summary.

.TP
\fB-v\fR, \fB--version\fR
Show the program version.

.TP
\fB-c\fR, \fB--config-file\fR
Configuration File.  Specify a configuration file to use.  
The program will read the default configuration file and then this 
configuration file.  See \fIapt.conf\fR(5) for syntax information.      

.TP
\fB-o\fR, \fB--option\fR
Set a Configuration Option.  This will set an arbitrary configuration 
option.  The syntax is \fB-o Foo::Bar=bar\fR.

.SH "SEE ALSO"
\fIapt.conf\fR(5)

.SH "DIAGNOSTICS"
\fBapt-config\fR returns zero on normal operation, decimal 100 on error.

.SH "BUGS"
Reporting bugs in APT-RPM is best done in the APT-RPM mailinglist at
http://apt-rpm.org/mailinglist.shtml.

.SH "AUTHOR"
Maintainer and contributor information can be found in the credits page
http://apt-rpm.org/about.shtml of APT-RPM.


ChangeLog
------------------------------------------------------------------------
r166 | pmatilai | 2006-05-22 18:11:09 +0300 (Mon, 22 May 2006) | 2 lines

- preparing for lorg3.1

------------------------------------------------------------------------
r165 | pmatilai | 2006-05-22 18:02:03 +0300 (Mon, 22 May 2006) | 2 lines

- updates to German translation (Ralf)

------------------------------------------------------------------------
r164 | pmatilai | 2006-05-17 20:10:37 +0300 (Wed, 17 May 2006) | 3 lines

- bump cache minor version number to force full cache rebuild after upgrade 
  from previous versions

------------------------------------------------------------------------
r163 | pmatilai | 2006-05-17 20:08:36 +0300 (Wed, 17 May 2006) | 2 lines

- remove unused rpm::multiarch optionhash

------------------------------------------------------------------------
r162 | pmatilai | 2006-05-17 08:53:55 +0300 (Wed, 17 May 2006) | 2 lines

- set HOME in apt-cron, otherwise signed repositories wont work from apt-cron 

------------------------------------------------------------------------
r161 | pmatilai | 2006-05-16 17:32:09 +0300 (Tue, 16 May 2006) | 2 lines

- fix off-by-one in filelists handling

------------------------------------------------------------------------
r160 | pmatilai | 2006-05-15 19:02:20 +0300 (Mon, 15 May 2006) | 9 lines

- In rpm < 4.2.1 zero epoch isn't the same as non-existent epoch, we can't
  but blindly assume a patched createrepo not adding false zero epochs was
  used and treat everything as is. With newer rpm's we can deal with the
  zero-epoch silliness but they need to be treated uniformly everywhere
  to keep apt's version bookkeeping in order, so we remove them. The other
  alternative of *adding* zero epochs everywhere is slightly more complex,
  looks uglier and causes some script breakage (although the scripts could
  be considered broken as well). 

------------------------------------------------------------------------
r159 | pmatilai | 2006-05-10 19:39:03 +0300 (Wed, 10 May 2006) | 2 lines

- remove bunch of unnecessary rpmlib includes

------------------------------------------------------------------------
r158 | pmatilai | 2006-05-10 18:39:15 +0300 (Wed, 10 May 2006) | 3 lines

- fix type-punning errors using temporary variables in rpmhandler + genpkglist
- fix "might be used uninitialized" complaint in apt-shell

------------------------------------------------------------------------
r157 | pmatilai | 2006-05-09 17:15:18 +0300 (Tue, 09 May 2006) | 2 lines

- add apt-cron from FE package to contrib

------------------------------------------------------------------------
r156 | pmatilai | 2006-05-08 23:55:20 +0300 (Mon, 08 May 2006) | 3 lines

- ignore gpg-pubkey "packages" unconditionally on rpm >= 4.1, they can't
  be handled with apt at all so they shouldn't be visible either

------------------------------------------------------------------------
r155 | pmatilai | 2006-05-08 23:44:22 +0300 (Mon, 08 May 2006) | 2 lines

- fix default rpm::source::build-command value to match reality

------------------------------------------------------------------------
r154 | pmatilai | 2006-05-08 23:38:26 +0300 (Mon, 08 May 2006) | 2 lines

- fix the default value of apt::cache::allnames to act like intended

------------------------------------------------------------------------
r153 | pmatilai | 2006-05-08 22:01:19 +0300 (Mon, 08 May 2006) | 4 lines

- use iterator in repomd handler offset handling to make it behave as the 
  handler interface requires thus fixing an off-by-one bug causing the first 
  package of repository to go ignored

------------------------------------------------------------------------
r152 | pmatilai | 2006-05-08 17:48:12 +0300 (Mon, 08 May 2006) | 4 lines

- replace the questionable magic to conditionally install headers in 
  buildlib/Makefile.am with AC_CONFIG_LINK'ed files, get rid of 
  buildlib/Makefile* (Ralf)

------------------------------------------------------------------------
r151 | pmatilai | 2006-05-08 09:44:16 +0300 (Mon, 08 May 2006) | 2 lines

- cleanup + simplify buildlib/tools.m4 handling (Ralf)

------------------------------------------------------------------------
r150 | pmatilai | 2006-05-07 12:33:49 +0300 (Sun, 07 May 2006) | 6 lines

- When stepping through source records skip binaries
- Move CreateSrcParser() to rpmRepomdIndex class from rpmRepomdSrcIndex
-> Mixed binary + source repomd suddenly works. repomd-src isn't strictly
needed anymore but should be used for efficiency reasons when using pure
source repositories.

------------------------------------------------------------------------
r149 | pmatilai | 2006-05-05 20:24:04 +0300 (Fri, 05 May 2006) | 2 lines

- resurrect some gettext-upgrade victims :)

------------------------------------------------------------------------
r148 | pmatilai | 2006-05-05 14:01:27 +0300 (Fri, 05 May 2006) | 2 lines

- honor quiet mode in rpm progress reporting (Daisuke Suzuki)

------------------------------------------------------------------------
r147 | pmatilai | 2006-05-05 13:54:36 +0300 (Fri, 05 May 2006) | 2 lines

- remove leftover files from gettext update + add some missing bits (Ralf)

------------------------------------------------------------------------
r146 | pmatilai | 2006-05-05 09:50:56 +0300 (Fri, 05 May 2006) | 2 lines

- actually remove intl/ directory + the zero-sized files

------------------------------------------------------------------------
r145 | pmatilai | 2006-05-05 09:22:10 +0300 (Fri, 05 May 2006) | 2 lines

- update to gettext 0.14.5, remove internal intl (Ralf)

------------------------------------------------------------------------
r144 | pmatilai | 2006-05-04 23:00:20 +0300 (Thu, 04 May 2006) | 3 lines

- kill one remaining extra e:v-r logic, use handler EVR instead
- remove a commented leftover Epoch() definition from rpmhandler

------------------------------------------------------------------------

cmdline/apt-config.cc
// -*- mode: c++; mode: fold -*-
// Description								/*{{{*/
// $Id: apt-config.cc,v 1.5 2003/01/29 18:43:48 niemeyer Exp $
/* ######################################################################
   
   APT Config - Program to manipulate APT configuration files
   
   This program will parse a config file and then do something with it.
   
   Commands:
     shell - Shell mode. After this a series of word pairs should occure.
             The first is the environment var to set and the second is
             the key to set it from. Use like: 
 eval `apt-config shell QMode apt::QMode`
   
   ##################################################################### */
									/*}}}*/
// Include Files							/*{{{*/
#include <apt-pkg/cmndline.h>
#include <apt-pkg/error.h>
#include <apt-pkg/init.h>
#include <apt-pkg/strutl.h>

#include <config.h>
#include <apti18n.h>

// CNC:2003-02-14 - apti18n.h includes libintl.h which includes locale.h,
// 		    as reported by Radu Greab.
//#include <locale.h>
#include <iostream>
#include <string>
									/*}}}*/
using namespace std;

// DoShell - Handle the shell command					/*{{{*/
// ---------------------------------------------------------------------
/* */
bool DoShell(CommandLine &CmdL)
{
   for (const char **I = CmdL.FileList + 1; *I != 0; I += 2)
   {
      if (I[1] == 0 || strlen(I[1]) == 0)
	 return _error->Error(_("Arguments not in pairs"));

      string key = I[1];
      if (key.end()[-1] == '/') // old directory format
	 key.append("d");

      if (_config->ExistsAny(key.c_str()))
	 cout << *I << "='" << 
	         SubstVar(_config->FindAny(key.c_str()),"'","'\\''") << '\'' << endl;
      
   }
   
   return true;
}
									/*}}}*/
// DoDump - Dump the configuration space				/*{{{*/
// ---------------------------------------------------------------------
/* */
bool DoDump(CommandLine &CmdL)
{
   _config->Dump(cout);
   return true;
}
									/*}}}*/
// ShowHelp - Show the help screen					/*{{{*/
// ---------------------------------------------------------------------
/* */
int ShowHelp()
{
   ioprintf(cout,_("%s %s for %s %s compiled on %s %s\n"),PACKAGE,VERSION,
	    COMMON_OS,COMMON_CPU,__DATE__,__TIME__);
   if (_config->FindB("version") == true)
      return 0;
   
   cout <<
    _("Usage: apt-config [options] command\n"
      "\n"
      "apt-config is a simple tool to read the APT config file\n"
      "\n"
      "Commands:\n"
      "   shell - Shell mode\n"
      "   dump - Show the configuration\n"
      "\n"
      "Options:\n"
      "  -h   This help text.\n" 
      "  -c=? Read this configuration file\n" 
      "  -o=? Set an arbitary configuration option, eg -o dir::cache=/tmp\n");
   return 0;
}
									/*}}}*/

int main(int argc,const char *argv[])
{
   CommandLine::Args Args[] = {
      {'h',"help","help",0},
      {'v',"version","version",0},
      {'c',"config-file",0,CommandLine::ConfigFile},
      {'o',"option",0,CommandLine::ArbItem},
      {0,0,0,0}};
   CommandLine::Dispatch Cmds[] = {{"shell",&DoShell},
                                   {"dump",&DoDump},
                                   {0,0}};

   // Set up gettext support
   setlocale(LC_ALL,"");
   textdomain(PACKAGE);

   // Parse the command line and initialize the package library
   CommandLine CmdL(Args,_config);
   if (pkgInitConfig(*_config) == false ||
       CmdL.Parse(argc,argv) == false ||
       pkgInitSystem(*_config,_system) == false)
   {
      _error->DumpErrors();
      return 100;
   }

   // See if the help should be shown
   if (_config->FindB("help") == true ||
       CmdL.FileSize() == 0)
      return ShowHelp();

   // Match the operation
   CmdL.DispatchArg(Cmds);
   
   // Print any errors or warnings found during parsing
   if (_error->empty() == false)
   {
      bool Errors = _error->PendingError();
      _error->DumpErrors();
      return Errors == true?100:0;
   }
   
   return 0;
}
